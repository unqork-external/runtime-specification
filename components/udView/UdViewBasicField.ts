import { Description, Optional } from '@tsed/schema'

import { Field } from '../../component-composition'

export class UdViewBasicField extends Field {
  @Optional()
  @Description('Accessible label that describes the view for screen readers.')
  ariaLabel?: string

  @Optional()
  @Description('Accessible ARIA role for assistive technology tools based.')
  ariaRole?: string

  @Optional()
  @Description('Allows customization of the class HTML parameter for the root element of the component')
  customClass?: string
}
