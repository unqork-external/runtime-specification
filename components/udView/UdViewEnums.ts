export enum UdViewAppearance {
  ASIDE = 'aside',
  BLOCK = 'block',
  FOOTER = 'footer',
  HEADER = 'header',
  INLINE = 'inline',
  LIST_ITEM = 'list-item',
  MAIN = 'main',
  NAVIGATION = 'navigation',
  NONE = 'none',
  ORDERED_LIST = 'ordered-list',
  SECTION = 'section',
  UNORDERED_LIST = 'unordered-list',
}
