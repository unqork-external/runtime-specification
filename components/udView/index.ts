export * from './udView.definition'
export * from './udViewTargets.enum'
