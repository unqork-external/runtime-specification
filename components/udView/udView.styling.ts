import { Description, Property } from '@tsed/schema'

import { UdViewTargets } from './udViewTargets.enum'
import { Css } from '../../../decorators/schema/css.decorator'
import { type CssClassOrProps } from '../../styling/cssClassOrProps.type'
import { StylingModel } from '../../styling/styling.model'
import { TargetsModel } from '../../viewTargets/targets.model'

export class UdViewStyleTargets implements TargetsModel<UdViewTargets> {
  @Css()
  @Description('Root element styles.')
  [UdViewTargets.root]: CssClassOrProps
}

export class UdViewStyling extends StylingModel<UdViewStyleTargets> {
  @Property(UdViewStyleTargets)
  declare targets: UdViewStyleTargets
}
