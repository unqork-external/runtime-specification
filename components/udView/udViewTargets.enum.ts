export enum UdViewTargets {
  root = 'root',
}
