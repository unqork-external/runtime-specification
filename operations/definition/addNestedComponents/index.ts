export * from './addNestedComponents.model'
export * from './addNestedComponents.options'
