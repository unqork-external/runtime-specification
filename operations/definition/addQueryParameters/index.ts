export * from './addQueryParameters.model'
export * from './addQueryParameters.options'
