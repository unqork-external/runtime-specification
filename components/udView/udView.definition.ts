import {
  CollectionOf,
  Default,
  Description,
  DiscriminatorValue,
  Enum,
  Optional,
  Property,
  Required,
} from '@tsed/schema'

import { UdViewStyling } from './udView.styling'
import { UdViewBasicField } from './UdViewBasicField'
import { UdViewAppearance } from './UdViewEnums'
import { UdViewTargets } from './udViewTargets.enum'
import { ViewTargets } from '../../../decorators/viewTargets/viewTargets.decorator'
import { BaseComponentDefinition } from '../../base-component-interface/base.component.definition'
import { Display } from '../../component-composition/display/component.display'
import { StandardArrayNestable } from '../../nestables'
import type { SignalTargets } from '../../signals'

@DiscriminatorValue('udView')
@ViewTargets(UdViewTargets)
export class UdViewComponentDefinition extends BaseComponentDefinition {
  @Required()
  type: 'udView' = 'udView' as const

  @Default(UdViewAppearance.BLOCK)
  @Enum(UdViewAppearance)
  @Description('Appearance represents the intended visual style and role for the view.')
  appearance?: UdViewAppearance

  @Property(Display)
  @Description('Display settings')
  display: Display = new Display()

  @Property(UdViewBasicField)
  @Description('Field settings')
  field: UdViewBasicField = new UdViewBasicField()

  @CollectionOf(BaseComponentDefinition)
  components?: BaseComponentDefinition[]

  @Optional()
  declare signals?: SignalTargets<UdViewTargets>

  @Optional()
  @Description('Map of customized CSS styling for specific targets.')
  @Property(UdViewStyling)
  declare styling?: UdViewStyling

  @Description('Standard nestable implementation')
  nestables = { childIds: new StandardArrayNestable() }
}

export class UdViewComponentState extends UdViewComponentDefinition {
  @Required()
  @Description('Child ids for all nestable references within the component.')
  childIds: string[] = []
}
