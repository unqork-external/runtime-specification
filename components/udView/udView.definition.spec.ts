import { UdViewComponentDefinition } from './udView.definition'
import { generateSchemaAndValidate } from '../../../utilities'

describe('UD View Contract Tests', function () {
  it('should ensure the contract is clear', function () {
    const { schema } = generateSchemaAndValidate(UdViewComponentDefinition)

    expect(schema.required).toMatchSnapshot()
    expect(schema.properties).toMatchSnapshot()
    expect(schema.definitions).toMatchSnapshot()
  })
})
