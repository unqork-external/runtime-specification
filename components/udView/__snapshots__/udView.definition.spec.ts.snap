// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UD View Contract Tests should ensure the contract is clear 1`] = `
[
  "key",
  "type",
]
`;

exports[`UD View Contract Tests should ensure the contract is clear 2`] = `
{
  "appearance": {
    "default": "block",
    "description": "Appearance represents the intended visual style and role for the view.",
    "enum": [
      "aside",
      "block",
      "footer",
      "header",
      "inline",
      "list-item",
      "main",
      "navigation",
      "none",
      "ordered-list",
      "section",
      "unordered-list",
    ],
    "type": "string",
  },
  "childIds": {
    "description": "Child IDs for nestable references",
    "type": "array",
  },
  "components": {
    "items": {
      "$ref": "#/definitions/BaseComponentDefinition",
    },
    "type": "array",
  },
  "direction": {
    "default": "vertical",
    "description": "Direction represents how the view children should be disposed.",
    "enum": [
      "horizontal",
      "default",
      "vertical",
    ],
    "type": "string",
  },
  "display": {
    "$ref": "#/definitions/Display",
  },
  "draggable": {
    "$ref": "#/definitions/UdViewDraggable",
  },
  "droppable": {
    "$ref": "#/definitions/UdViewDroppable",
  },
  "executable": {
    "default": true,
    "description": "Defines whether or not the object will respond to events or fire operations. When \`true\`, the object will perform and behave as defined in its spec. When \`false\`, the object will no longer be responsive to the system.",
    "type": "boolean",
  },
  "field": {
    "$ref": "#/definitions/UdViewBasicField",
  },
  "inputs": {
    "items": {
      "$ref": "#/definitions/InputRef",
    },
    "type": "array",
  },
  "key": {
    "description": "'key' is the user-defined, unique identifier for an object within a module. It doesn't have to be unique across different modules. Objects within the runtime use keys to target other objects.",
    "examples": [
      "I am the key",
      "bestKeyEver",
    ],
    "minLength": 1,
    "type": "string",
  },
  "nestables": {
    "description": "Standard nestable implementation",
    "type": "object",
  },
  "package": {
    "$ref": "#/definitions/PackageModel",
  },
  "signals": {
    "description": "An object which contains all of the signals for the component. The signals object can have any number of target keys as defined in the component definition which each contain an array of Signals.",
    "examples": [
      {
        "args": {
          "inputs": {
            "propertyPath": "value",
            "required": true,
            "targetKey": "firstName",
          },
        },
        "id": "firstNameWatcher",
        "operations": [
          {
            "options": {
              "logType": "warn",
              "thingToPrint": "firstNameWatcher operations fired: value watcher triggered",
            },
            "type": "PRINTER",
          },
        ],
        "type": "onWatch",
      },
      {
        "id": "textfieldOnBlur",
        "operations": [
          {
            "options": {
              "logType": "warn",
              "thingToPrint": "textfield received blur event from DOM",
            },
            "type": "PRINTER",
          },
        ],
        "type": "onBlur",
      },
      "<Example> Prompt: "I want a textfield that will trigger the root execute signal on myValidationComponent when the textfield input receives a blur event." Response: [{"key":"yourGeneratedTextfield","type":"textfield","signals":{"input":[{"id":"textfieldOnBlur","operations":[{"type":"EXECUTE","options":{"targetKey":"myValidationComponent","signalTarget":"root","shouldExecute":true}}],"type":"onBlur"}]}}] </Example>",
    ],
    "type": "object",
  },
  "styling": {
    "$ref": "#/definitions/StylingModel",
  },
  "type": {
    "const": "udView",
    "description": "Name of the type of component.",
    "examples": [
      "udView",
    ],
    "minLength": 1,
    "type": "string",
  },
  "validation": {
    "$ref": "#/definitions/Validation",
  },
  "value": {
    "$ref": "#/definitions/UdViewDroppableValue",
    "description": "Current value of the component",
    "type": [
      "null",
      "integer",
      "number",
      "string",
      "boolean",
      "array",
      "object",
    ],
  },
}
`;

exports[`UD View Contract Tests should ensure the contract is clear 3`] = `
{
  "BaseComponentDefinition": {
    "description": "Base Component Definition is the fundamental interface any component that wishes to leverage the runtime must follow. It is the agreed upon contract between the Runtime and anything that should be recognized as a component.",
    "properties": {
      "executable": {
        "default": true,
        "description": "Defines whether or not the object will respond to events or fire operations. When \`true\`, the object will perform and behave as defined in its spec. When \`false\`, the object will no longer be responsive to the system.",
        "type": "boolean",
      },
      "inputs": {
        "items": {
          "$ref": "#/definitions/InputRef",
        },
        "type": "array",
      },
      "key": {
        "description": "'key' is the user-defined, unique identifier for an object within a module. It doesn't have to be unique across different modules. Objects within the runtime use keys to target other objects.",
        "examples": [
          "I am the key",
          "bestKeyEver",
        ],
        "minLength": 1,
        "type": "string",
      },
      "package": {
        "$ref": "#/definitions/PackageModel",
      },
      "signals": {
        "description": "An object which contains all of the signals for the component. The signals object can have any number of target keys as defined in the component definition which each contain an array of Signals.",
        "examples": [
          {
            "args": {
              "inputs": {
                "propertyPath": "value",
                "required": true,
                "targetKey": "firstName",
              },
            },
            "id": "firstNameWatcher",
            "operations": [
              {
                "options": {
                  "logType": "warn",
                  "thingToPrint": "firstNameWatcher operations fired: value watcher triggered",
                },
                "type": "PRINTER",
              },
            ],
            "type": "onWatch",
          },
          {
            "id": "textfieldOnBlur",
            "operations": [
              {
                "options": {
                  "logType": "warn",
                  "thingToPrint": "textfield received blur event from DOM",
                },
                "type": "PRINTER",
              },
            ],
            "type": "onBlur",
          },
          "<Example> Prompt: "I want a textfield that will trigger the root execute signal on myValidationComponent when the textfield input receives a blur event." Response: [{"key":"yourGeneratedTextfield","type":"textfield","signals":{"input":[{"id":"textfieldOnBlur","operations":[{"type":"EXECUTE","options":{"targetKey":"myValidationComponent","signalTarget":"root","shouldExecute":true}}],"type":"onBlur"}]}}] </Example>",
        ],
        "type": "object",
      },
      "styling": {
        "$ref": "#/definitions/StylingModel",
      },
      "type": {
        "description": "Name of the type of component.",
        "minLength": 1,
        "type": "string",
      },
      "validation": {
        "$ref": "#/definitions/Validation",
      },
      "value": {
        "description": "Current value of the component",
        "type": [
          "null",
          "integer",
          "number",
          "string",
          "boolean",
          "array",
          "object",
        ],
      },
    },
    "required": [
      "key",
      "type",
    ],
    "type": "object",
  },
  "Display": {
    "description": "Basic UI settings related to how a component may be displayed and interacted with. 'Hidden' corresponds to visibility. 'Interactive' corresponds to the end-user's ability to interact with input components. 'Read Only View' corresponds to the component being rendered with an input element or as text.",
    "properties": {
      "hidden": {
        "default": false,
        "description": "Setting it to \`true\` hides the component from the view, while \`false\` shows it. Regardless of this setting, the component is always visible in the Module Editor. A hidden component can still receive information from other components.",
        "type": "boolean",
      },
      "interactive": {
        "default": true,
        "description": "When \`false\`, it is not possible to interact with or edit the input field. Its appearance is modified (grayed-out by default) to show its state, the HTML 'disabled' attribute is added to the field, and when hovering over it, the cursor also changes to alert of the disabled status. However, the component can still receive information from other components when it's not interactive.",
        "type": "boolean",
      },
      "readOnlyView": {
        "default": false,
        "description": "Defines how the component will be displayed. When \`false\`, the component is displayed as an input field. When \`true\`, the component will not accept any input from the user, and its value will be rendered inline as text and not as an input component. If no value is present, the word 'None' is rendered. If a prefix or suffix is set for the component, they also appear as inline text.",
        "type": "boolean",
      },
    },
    "type": "object",
  },
  "InputRef": {
    "description": "Objects may define other targets that they refer to. This becomes a link for operation execution within the Runtime.",
    "properties": {
      "alias": {
        "type": "string",
      },
      "propertyPath": {
        "type": "string",
      },
      "required": {
        "type": "boolean",
      },
      "silent": {
        "type": "boolean",
      },
      "targetKey": {
        "minLength": 1,
        "type": "string",
      },
      "type": {
        "type": "string",
      },
    },
    "required": [
      "targetKey",
    ],
    "type": "object",
  },
  "Label": {
    "properties": {
      "text": {
        "description": "Text displayed as the label for the field",
        "type": "string",
      },
    },
    "type": "object",
  },
  "PackageModel": {
    "description": "The PackageModel describes how the Vega Runtime will resolve dependencies",
    "properties": {
      "type": {
        "description": "Type of package",
        "type": "string",
      },
      "url": {
        "description": "URL of package",
        "type": "string",
      },
      "version": {
        "description": "Version of package",
        "type": "string",
      },
    },
    "type": "object",
  },
  "StylingModel": {
    "description": "Holds information for targeted styling of the component.",
    "properties": {
      "targets": {
        "type": "object",
      },
    },
    "type": "object",
  },
  "UdViewBasicField": {
    "properties": {
      "customClass": {
        "description": "Allows customization of the class HTML parameter for the root element of the component",
        "type": "string",
      },
      "label": {
        "$ref": "#/definitions/Label",
      },
    },
    "type": "object",
  },
  "UdViewDraggable": {
    "properties": {
      "dragData": {
        "description": "The data this object transmits when being dropped",
        "type": "string",
      },
      "dragGroup": {
        "description": "The drag group this component belongs to",
        "type": "string",
      },
      "isDraggable": {
        "description": "Indicates a component is draggable",
        "type": "boolean",
      },
    },
    "type": "object",
  },
  "UdViewDroppable": {
    "properties": {
      "allowedDragGroups": {
        "description": "Array of strings for the drag groups that can be dropped in this component",
        "type": "array",
      },
      "dropTrigger": {
        "description": "Component key to trigger when a component is dropped",
        "type": "string",
      },
      "emptyText": {
        "description": "Text to display when the component has no children",
        "type": "string",
      },
      "isDroppable": {
        "description": "Indicates a component is droppable",
        "type": "boolean",
      },
      "locationId": {
        "description": "Location metadata to be sent to the parent droppable component in case a droppable child receives a drop",
        "type": "string",
      },
      "placementText": {
        "description": "Text to display in the location that will accept a component while its being dragged over this component",
        "type": "string",
      },
    },
    "type": "object",
  },
  "UdViewDroppableValue": {
    "properties": {
      "droppedDragData": {
        "description": "The data from the dropped item",
        "type": "string",
      },
      "droppedDragGroup": {
        "description": "The drag group of the dropped item",
        "type": "string",
      },
      "droppedIndex": {
        "description": "Index within the droppable component where the drop happened, -1 if above the first element",
        "type": "number",
      },
      "droppedLocationId": {
        "description": "Location metadata from the droppable component highest on the DOM tree that received the drop",
        "type": "string",
      },
    },
    "type": "object",
  },
  "Validation": {
    "description": "Holds the validation rules and the most current validation results for the component",
    "properties": {
      "isDirty": {
        "description": "Whether or not field has had user interaction. It will be set to true on user interaction.",
        "type": "boolean",
      },
      "result": {
        "$ref": "#/definitions/ValidationResult",
      },
      "rules": {
        "description": "The set of rules to restrict or define the accepted data for the current component. These are the rules to run for validation of the data input.",
        "examples": [
          "You can use it to define the max length of characters in an input field with a custom error message. rules: { required: { type: ValidationRuleTypes.MAX_LENGTH, validation: { maxLength: 5, }, errorMessage: 'Max length is 5', }, },",
        ],
        "type": "object",
      },
    },
    "type": "object",
  },
  "ValidationError": {
    "properties": {
      "message": {
        "description": "The error message",
        "type": "string",
      },
      "type": {
        "description": "Type of the error",
        "enum": [
          "required",
          "maxLength",
          "minLength",
          "arrayMinLength",
          "arrayMaxLength",
          "pattern",
          "min",
          "max",
          "mask",
          "number",
          "dateinput",
          "isBeforeDate",
          "isAfterDate",
          "customError",
          "isEmail",
        ],
        "minLength": 1,
        "type": "string",
      },
    },
    "required": [
      "type",
    ],
    "type": "object",
  },
  "ValidationResult": {
    "properties": {
      "errors": {
        "items": {
          "$ref": "#/definitions/ValidationError",
        },
        "type": "array",
      },
      "isValid": {
        "type": [
          "null",
          "boolean",
        ],
      },
    },
    "type": "object",
  },
}
`;
